packages/modal/src/Modal.tsx

```diff
@@ -0,0 +1,253 @@
+  size?: ModalSize;
+
+  /**
+   * Callback to change the active state of the Modal.
+   *
+   */
+  setActive?: (
+    active: boolean,
+  ) => void | React.Dispatch<SetStateAction<boolean>>;
+
+  /**
+   * Callback to determine whether or not Modal should close when user tries to close it.
+   *
+   */
+  shouldClose?: () => boolean;
+
+  /**
+   * className applied to root overlay div.
+   *
+   */
+  className?: string;
+}
+
+/**
+ * # Modal
+ *
+ *  Modals place content on top of main window.
+ *
+```
+<Modal
+  active
+  size="large"
+  setActive={setActive}
+  shouldClose={() => console.log('Modal is closing now!')}
+  >  
+  Modal content!
+</Modal>
+```
+ * @param props.active Boolean to describe whether or not Modal is active.
+ * @param props.size String to determine size of Modal. ['xxsmall', 'xsmall', 'small', 'default', 'large', 'xlarge']
+ * @param props.setActive Callback to change the active state of Modal.
+ * @param props.children Content to appear inside of Modal container.
+ * @param props.shouldClose Callback to determine whether or not Modal should close when user tries to close it.
+ * @param props.className className applied to overlay div.
+ *
+ */
+function Modal({
+  active = false,
+  size = ModalSize.Default,
+  setActive,
+  children,
+  shouldClose,
+  className,
+  ...rest
+}: ModalProps) {
+  if (!active) {
+    return null;
+  }
+
+  const contentRef = useRef<HTMLDivElement>(null);
+
+  const handleClose = () => {
+    // Don't close modal if shouldClose returns false or no setActive callback is passed
+    if (!setActive || (shouldClose && !shouldClose())) {
+      return;
+    }
+
+    setActive(false);
+  };
+
+  const handleDocumentClick = (e: React.SyntheticEvent) => {
+    if (!contentRef.current) {
+      return;
+    }
+
+    if (!contentRef.current.contains(e.target as Node)) {
+      handleClose();
+    }
+  };
+
+  const handleEscape = useCallback((e: KeyboardEvent) => {
+    if (e.keyCode === escapeKey) {
+      handleClose();
+    }
+  }, []);
+
+  useEventListener('keydown', handleEscape, {
+    options: { once: true },
+  });
+
+  injectGlobal(`
+    html, body {
+      overflow: hidden;
+    }
+
+    body {
+      position: relative;
+    }
+  `);
+
+  return (
+    <Portal>
+      <div {...rest} onClick={handleDocumentClick} className={scrollContainer}>
+        <div
+          className={cx(modalContentStyle, modalSizes[size], className)}
+          tabIndex={-1}
+          ref={contentRef}
+        >
+          <Icon
+            glyph="X"
+            fill={'#5D6C74'}
+            size={Size.Large}
+            onClick={handleClose}
+            className={closeButton}
+            data-dismiss="modal"
+            aria-hidden="true"
+          />
+
+          {children}
+        </div>
+      </div>
+    </Portal>
+  );
+}
+
+Modal.displayName = 'Modal';
+
+Modal.propTypes = {
+  active: PropTypes.bool,
+  size: PropTypes.string,
+  children: PropTypes.node,
+  shouldClose: PropTypes.func,
```

Comment:
Should this be `setActive`?

Reply:
Oh I was wrong, `shouldClose` is a real thing. But also we should add a propType for `setActive`